omnigibson/utils/usd_utils.py

```diff
@@ -568,12 +748,291 @@ def get_world_pose_with_scale(cls, prim_path):
 
 
+class BatchControlViewAPIImpl:
+    """
+    A centralized view that allows for reading and writing to an ArticulationView that covers all
+    controllable objects in the scene. This is used to avoid the overhead of reading from many views
+    for each robot in each physics step, a source of significant overhead.
+    """
+
+    def __init__(self, pattern):
+        # The prim path pattern that will be passed into the view
+        self._pattern = pattern
+
+        # The unified ArticulationView used to access all of the controllable objects in the scene.
+        self._view = None
+
+        # Cache for all of the view functions' return values within the same simulation step.
+        # Keyed by function name without get_, the value is the return value of the function.
+        self._read_cache = {}
+
+        # Cache for all of the view functions' write values within the same simulation step.
+        # Keyed by the function name without set_, the value is the set of indices that need to be updated.
+        self._write_idx_cache = collections.defaultdict(set)
+
+        # Mapping from prim path to index in the view.
+        self._idx = {}
+
+        # Mapping from prim idx to a dict that maps link name to link index in the view.
+        self._link_idx = {}
+
+    def clear(self):
+        self._read_cache = {}
+        self._write_idx_cache = collections.defaultdict(set)
+
+    def flush_control(self):
+        if "dof_position_targets" in self._write_idx_cache:
+            pos_indices = np.array(sorted(self._write_idx_cache["dof_position_targets"]))
+            pos_targets = self._read_cache["dof_position_targets"]
+            self._view.set_dof_position_targets(pos_targets, np.array(pos_indices))
+
+        if "dof_velocity_targets" in self._write_idx_cache:
+            vel_indices = np.array(sorted(self._write_idx_cache["dof_velocity_targets"]))
+            vel_targets = self._read_cache["dof_velocity_targets"]
+            self._view.set_dof_velocity_targets(vel_targets, np.array(vel_indices))
+
+        if "dof_actuation_forces" in self._write_idx_cache:
+            eff_indices = np.array(sorted(self._write_idx_cache["dof_actuation_forces"]))
+            eff_targets = self._read_cache["dof_actuation_forces"]
+            self._view.set_dof_actuation_forces(eff_targets, np.array(eff_indices))
+
+    def initialize_view(self):
+        # First, get all of the controllable objects in the scene (avoiding circular import)
+        from omnigibson.objects.controllable_object import ControllableObject
+
+        controllable_objects = [
+            obj for scene in og.sim.scenes for obj in scene.objects if isinstance(obj, ControllableObject)
+        ]
+
+        # This only works if the root link is called base_link for every controllable object, so assert that
+        assert all(
+            co.root_link.prim_path.endswith("/base_link") for co in controllable_objects
```

Comment:
this is pretty hacky, but idk how we can avoid this

Reply:
I wonder if we can rename a prim so we have a hardcoded string prefix that we can set globally across all controllable objects

Reply:
not sure what this means

Reply:
TODO: in controllable object, after initialize, check root link is named as base_link

Reply:
@cgokmen is there any way to have a single string regex expression to capture any number of root link names? e.g.: `'root_link | base_link | panda_base | ...'`?

Reply:
Oops sorry I wrote a response to this but forgot to post. Long story short is that these are glob patterns that only support * in the path. You can technically do what you said above by creating a separate view for each of the possible root link names, but each fetch adds latency and also the code gets quite complex so I would prefer not to do it.